Split --uktdd-body-position-try-fallbacks into two fixed slots (#404)
acusti/uikit: .changeset/dropdown-fill-fallbacks-within-evaluation-limit.md

'@acusti/dropdown': minor
---

Budget the last-resort fill placements to the `position-try` evaluation
limit so a too-tall body flips and scrolls instead of opening off-screen
(fixes #399)

The last-resort fills a too-tall body falls back to have to fit within the
`position-try` evaluation limit, or an engine silently drops one and the

> body opens toward a viewport edge instead of flipping and scrolling. The

Author: i don’t follow the beginning of this sentence: “The last-resort fills a too-tall body falls back…”

seems like too many verbs? “fills” and “falls back”?

Author: Reworded in e5a75846 — now "A too-tall body falls back to last-resort fill placements, which have to fit within the `position-try` evaluation limit…" (drops the "fills … falls back to" verb pileup).

position, exactly the spec’s guaranteed floor of five options, so no engine
can silently drop a fill. The two author fallbacks are the single-axis
flips of the primary (a block flip and an inline flip); the both-axes
diagonal is dropped to stay within the floor. The cost of dropping the
diagonal is that a trigger crammed into the corner diagonal to the primary
opening direction lands in a fill rather than a fourth natural placement,
which for a corner-pinned trigger is a re-tune-the-primary case anyway.
Each fill pairs a single-side `position-area` (`block-end`/`block-start`,
spanning all inline tiles so it can never be rejected for horizontal
overflow) with `justify-self: anchor-center`, which centers the body over
the trigger and shifts it inward as needed to stay on-screen (the
explicitness matters: `span-all`’s default alignment centers without that
shift, so near a corner it would overflow and be rejected). A fill is only
rejected when its side is shorter than the body’s min-block-size floor —
what sends a cramped side’s fill to the roomier opposite side — and each
fill caps that floor at the worst-case larger side (half the viewport minus

> the trigger, the most a block-centered trigger can guarantee), so at least

Author: technically our math doesn’t include the trigger, right? just the gap?

Author: Good question, but the trigger *is* in it: the cap is `(100dvb - anchor-size(block)) / 2 - gap·2` — so it subtracts both the trigger (anchor-size) and the gap. The changeset actually omitted the gap. Rather than spell out the full formula in release notes, I simplified it to "roughly half the viewport" (e5a75846); the precise `(100dvb - anchor)/2` lives in the @position-try comment.

'@acusti/dropdown': minor
---

Budget the last-resort fill placements to the `position-try` evaluation
limit so a too-tall body flips and scrolls instead of opening off-screen
(fixes #399)

A too-tall body falls back to last-resort fill placements, which have to
fit within the `position-try` evaluation limit, or an engine silently drops
one and the body opens toward a viewport edge instead of flipping and
scrolling. The CSS anchor positioning spec lets engines cap the position
options list at an implementation-defined length with a floor of five — and
the list includes the element’s base position, so only four fallbacks past

> it are guaranteed. In practice Chromium is the tightest, evaluating six

Author: ```suggestion
options list at an implementation-defined length with a floor of five,
including the element’s base position, so only four fallbacks past
```

Author: Applied in 38351e75.

options (five fallbacks past the base) and silently ignoring the rest,
while Safari and Firefox evaluate many more.

The default budget is two author fallbacks plus two named fill options —
`--uktdd-fill-bottom` and `--uktdd-fill-top` — four fallbacks past the base
position, exactly the spec’s guaranteed floor of five options, so no engine

> can silently drop a fill. The two author fallbacks are the single-axis

Author: ```suggestion
The default budget is two author fallbacks plus two named fill options,
`--uktdd-fill-bottom` and `--uktdd-fill-top`, so no engine
can silently drop a fill. The two author fallbacks are the single-axis
```

Author: Applied in 38351e75.

flips of the primary (a block flip and an inline flip); the both-axes

> diagonal is dropped to stay within the floor. The cost of dropping the

Author: ```suggestion
flips of the primary (a block flip and an inline flip); the opposite
diagonal flip is dropped to stay within the floor. The cost of dropping the
```

Author: Applied in 38351e75.

diagonal is that a trigger crammed into the corner diagonal to the primary
opening direction lands in a fill rather than a fourth natural placement,

> which for a corner-pinned trigger is a re-tune-the-primary case anyway.

Author: ```suggestion
opening direction lands in a fill rather than a fourth natural placement.
```

Author: Applied in 38351e75.

Each fill pairs a single-side `position-area` (`block-end`/`block-start`,
spanning all inline tiles so it can never be rejected for horizontal
overflow) with `justify-self: anchor-center`, which centers the body over
the trigger and shifts it inward as needed to stay on-screen (the
explicitness matters: `span-all`’s default alignment centers without that
shift, so near a corner it would overflow and be rejected). A fill is only
rejected when its side is shorter than the body’s min-block-size floor —

> what sends a cramped side’s fill to the roomier opposite side — and each

Author: ```suggestion
rejected when its side is shorter than the body’s `min-block-size` floor, and each
```

Author: Applied in 38351e75.

> fill caps that floor at the worst-case larger side (roughly half the

Author: ```suggestion
fill caps that floor at the larger side (roughly half the
```

Author: Applied in 38351e75.

viewport, the most a block-centered trigger can guarantee), so at least one
of the pair always fits — for every trigger position, including corners and
a `--uktdd-body-min-height` above half the viewport, with no height cap

> needed. The pair is appended via the new `--uktdd-body-fill-fallbacks`

Author: ```suggestion
of the pair always fits, no additional height cap needed.
The pair is appended via the new `--uktdd-body-fill-fallbacks`
```

Author: Applied in 38351e75.

In the fill placements the body now spans the trigger (anchor-centered,

> shifted to fit) instead of keeping the primary placement’s edge alignment.

Author: is this somewhat obviated by this explanation above?
> …with `justify-self: anchor-center`, which centers the body over the trigger and shifts it inward as needed to stay on-screen

Author: Yes — cut the redundant "(anchor-centered, shifted to fit)" in e5a75846; the sentence now just makes the distinct point (fills span the trigger instead of keeping the primary's edge alignment).
